File: backend/src/services/aiInterpretation.js
/**
 * Module 2a — AI intake interpretation (Gemini).
 *
 * Turns messy free-text chief complaints into structured signals that feed
 * INTO the deterministic scoring engine as an additional input — this
 * module never sets severityLevel or severityScore itself. If the API key
 * is missing, the call errors, times out, or the response isn't valid JSON
 * matching the expected shape, we fail soft: return an "unavailable" result
 * so triage can proceed on keyword-only logic (module 2's own red-flag /
 * symptom keyword checks always run regardless of this module's outcome).
 */

const EXTRACTED_SIGNAL_VALUES = [
  "respiratory",
  "cardiac",
  "neurological",
  "allergic",
  "gi",
  "trauma",
  "infectious",
  "other",
];

const TIMEOUT_MS = 4000;
const MODEL_NAME = "gemini-1.5-flash";

function unavailableResult() {
  return {
    available: false,
    extractedSignals: [],
    ambiguityDetected: false,
    missingFieldSuggestions: [],
    narrative: "",
  };
}

function buildPrompt({ chiefComplaint, age, vitals }) {
  const vitalsSummary = vitals
    ? Object.entries(vitals)
        .filter(([, v]) => v != null)
        .map(([k, v]) => `${k}: ${v}`)
        .join(", ") || "none recorded"
    : "none recorded";

  return `You are a clinical intake text interpreter for an Emergency Department triage prototype.
You do NOT decide severity or triage level. You only extract structured signals from free text.

Patient age (years): ${age}
Recorded vitals: ${vitalsSummary}
Chief complaint (raw, patient/caregiver words): "${chiefComplaint}"

Return ONLY valid JSON, no markdown fences, no commentary, matching exactly this shape:
{
  "extractedSignals": ["respiratory"|"cardiac"|"neurological"|"allergic"|"gi"|"trauma"|"infectious"|"other", ...],
  "ambiguityDetected": true|false,
  "missingFieldSuggestions": ["short string", ...],
  "narrative": "one short plain-language clinical summary sentence"
}

Rules:
- extractedSignals: only include categories genuinely suggested by the text; may be empty array.
- ambiguityDetected: true if the complaint is vague, non-specific, or hard to map to a clear symptom category.
- missingFieldSuggestions: short suggestions of what additional info would help (e.g. "ask about onset time"), display-only.
- narrative: one short sentence, plain language, no diagnosis, no treatment advice.
- Output must be a single JSON object and nothing else.`;
}

function isValidShape(obj) {
  if (!obj || typeof obj !== "object") return false;
  if (!Array.isArray(obj.extractedSignals)) return false;
  if (!obj.extractedSignals.every((s) => EXTRACTED_SIGNAL_VALUES.includes(s))) return false;
  if (typeof obj.ambiguityDetected !== "boolean") return false;
  if (!Array.isArray(obj.missingFieldSuggestions)) return false;
  if (!obj.missingFieldSuggestions.every((s) => typeof s === "string")) return false;
  if (typeof obj.narrative !== "string") return false;
  return true;
}

function parseModelText(text) {
  if (!text) return null;
  // Strip common markdown code-fence wrapping defensively, in case the model
  // doesn't perfectly follow the "no fences" instruction.
  const cleaned = text.replace(/```json/gi, "").replace(/```/g, "").trim();
  try {
    const parsed = JSON.parse(cleaned);
    if (isValidShape(parsed)) return parsed;
    return null;
  } catch (e) {
    return null;
  }
}

async function withTimeout(promise, ms) {
  let timeoutHandle;
  const timeout = new Promise((_, reject) => {
    timeoutHandle = setTimeout(() => reject(new Error("AI_TIMEOUT")), ms);
  });
  try {
    const result = await Promise.race([promise, timeout]);
    clearTimeout(timeoutHandle);
    return result;
  } catch (err) {
    clearTimeout(timeoutHandle);
    throw err;
  }
}

/**
 * Interpret a patient's raw chief complaint via Gemini.
 * Always resolves (never rejects) — failures resolve to an "unavailable" result.
 */
async function interpretChiefComplaint({ chiefComplaint, age, vitals }) {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) {
    return unavailableResult();
  }

  try {
    // Lazy-require so the app still boots fine if the package isn't installed yet.
    const { GoogleGenerativeAI } = require("@google/generative-ai");
    const genAI = new GoogleGenerativeAI(apiKey);
    const model = genAI.getGenerativeModel({ model: MODEL_NAME });

    const prompt = buildPrompt({ chiefComplaint, age, vitals });
    const result = await withTimeout(model.generateContent(prompt), TIMEOUT_MS);
    const text = result?.response?.text ? result.response.text() : null;
    const parsed = parseModelText(text);

    if (!parsed) {
      return unavailableResult();
    }

    return {
      available: true,
      extractedSignals: parsed.extractedSignals,
      ambiguityDetected: parsed.ambiguityDetected,
      missingFieldSuggestions: parsed.missingFieldSuggestions,
      narrative: parsed.narrative,
    };
  } catch (err) {
    // Any failure (missing package, network error, timeout, bad response) —
    // fall back gracefully. Triage must never block on this call.
    return unavailableResult();
  }
}

module.exports = { interpretChiefComplaint, unavailableResult, EXTRACTED_SIGNAL_VALUES };

File: backend/app/ai_interpretation.py
import asyncio
import json
import re
from typing import List, Optional

from app.config import settings
from app.scoring.engine import RED_FLAG_KEYWORDS, SYMPTOM_KEYWORDS, VAGUE_PHRASES

_GEMINI_TIMEOUT_SECONDS = 4.0

_PROMPT_TEMPLATE = """You are assisting an Emergency Department intake system. You do NOT decide \
triage severity - a separate deterministic engine does that. Your only job is to interpret messy, \
free-text patient-reported complaints into structured signals.

Patient age: {age}
Available vitals: {vitals}
Chief complaint (verbatim): "{chief_complaint}"

Return ONLY valid JSON, no markdown fences, no commentary, matching exactly this shape:
{{
  "extractedSignals": ["respiratory"|"cardiac"|"neurological"|"allergic"|"gi"|"trauma"|"infectious"|"other", ...],
  "ambiguityDetected": true|false,
  "missingFieldSuggestions": ["short string", ...],
  "narrative": "one short plain-language clinical summary sentence"
}}
"""

_VALID_SIGNALS = {"respiratory", "cardiac", "neurological", "allergic", "gi", "trauma", "infectious", "other"}


def _keyword_fallback(chief_complaint: str) -> dict:
    """Deterministic keyword-only interpretation used whenever the AI call is unavailable."""
    complaint_lower = (chief_complaint or "").lower()
    signals: List[str] = []
    for category, keywords in SYMPTOM_KEYWORDS.items():
        if category == "other":
            continue
        if any(kw in complaint_lower for kw in keywords):
            signals.append(category)

    vague = any(p in complaint_lower for p in VAGUE_PHRASES)
    no_keyword_match = not signals and not any(kw in complaint_lower for kw in RED_FLAG_KEYWORDS)
    ambiguity_detected = bool(vague and no_keyword_match)

    return {
        "available": False,
        "extractedSignals": signals,
        "ambiguityDetected": ambiguity_detected,
        "missingFieldSuggestions": [],
        "narrative": "",
    }


def _parse_gemini_json(raw_text: str) -> Optional[dict]:
    text = raw_text.strip()
    fence_match = re.match(r"^```(?:json)?\s*(.*?)\s*```$", text, re.DOTALL)
    if fence_match:
        text = fence_match.group(1).strip()
    try:
        data = json.loads(text)
    except (json.JSONDecodeError, ValueError):
        return None

    if not isinstance(data, dict):
        return None

    signals = data.get("extractedSignals", [])
    if not isinstance(signals, list):
        return None
    signals = [s for s in signals if s in _VALID_SIGNALS]

    ambiguity = data.get("ambiguityDetected", False)
    if not isinstance(ambiguity, bool):
        return None

    suggestions = data.get("missingFieldSuggestions", [])
    if not isinstance(suggestions, list):
        suggestions = []

    narrative = data.get("narrative", "")
    if not isinstance(narrative, str):
        narrative = ""

    return {
        "available": True,
        "extractedSignals": signals,
        "ambiguityDetected": ambiguity,
        "missingFieldSuggestions": [str(s) for s in suggestions],
        "narrative": narrative,
    }


async def _call_gemini(prompt: str) -> Optional[str]:

    try:
        import google as genai
    except ImportError:
        return None

    def _sync_call() -> Optional[str]:
        genai.configure(api_key=settings.gemini_api_key)
        model = genai.GenerativeModel(settings.gemini_model)
        response = model.generate_content(prompt)
        return getattr(response, "text", None)

    try:
        return await asyncio.wait_for(asyncio.to_thread(_sync_call), timeout=_GEMINI_TIMEOUT_SECONDS)
    except Exception:
        return None


async def interpret_intake(age: float, chief_complaint: str, vitals: dict) -> dict:
    """
    Runs once per new patient, before scoring. Always returns a well-formed
    aiInterpretation dict, falling back to keyword-only interpretation on
    any failure (missing key, API error, timeout, malformed response).
    """
    if not settings.gemini_api_key:
        return _keyword_fallback(chief_complaint)

    prompt = _PROMPT_TEMPLATE.format(age=age, vitals=json.dumps(vitals), chief_complaint=chief_complaint)
    raw = await _call_gemini(prompt)
    if raw is None:
        return _keyword_fallback(chief_complaint)

    parsed = _parse_gemini_json(raw)
    if parsed is None:
        return _keyword_fallback(chief_complaint)

    return parsed
